File: Resources/TOI-Zero_68/A2/007/007.py
n, m = map(int, input().split())

adj = [[] for _ in range(n + 1)]
co = [0] * (n + 1)
dep = [0] * (n + 1)
par = list(range(n + 1))

a = [None] * (n + 1)
for i in range(1, n + 1):
    x, y = map(int, input().split())
    a[i] = (x, y)

for i in range(1, n + 1):
    for j in range(i + 1, n + 1):
        if a[j][1] > a[i][1]:
            break
        else:
            par[j] = i

for i in range(1, n + 1):
    if i != par[i]:
        adj[par[i]].append(i)

for _ in range(m):
    x = int(input())
    co[x] = 1

def dfs(u):
    for v in adj[u]:
        dep[v] = dep[u] + 1
        dfs(v)
        co[u] += co[v]

def dfs2(u, mx):
    if co[u] != mx:
        return -1, -1
    a = u
    b = dep[u]
    for v in adj[u]:
        x, y = dfs2(v, mx)
        if x == -1:
            continue
        if y > b:
            a = x
            b = y
    return a, b

ans = []
for i in range(1, n + 1):
    if par[i] == i:
        dfs(i)
        if co[i] == 0:
            continue
        a, _ = dfs2(i, co[i])
        ans.append(a)

print(len(ans))
ans.sort()
print(*ans)

File: Resources/Projects/Archive/2023/Competition/Programming/Contest/TOI-Zero_68/A2/019/TOI-Zero_68/A2/007/007.cpp
#include<bits/stdc++.h>
using namespace std;

#define ll long long

vector<int> adj[305];
int co[305],dep[305];

void dfs(int u){
	for(int i=0;i<adj[u].size();i++){
		int v=adj[u][i];
		dep[v]=dep[u]+1;
		dfs(v);
		co[u]+=co[v];
	}
}

pair<int,int> dfs2(int u,int mx){
	if(co[u]!=mx){
		return {-1,-1};
	}
	int a=u,b=dep[u];
	for(int i=0;i<adj[u].size();i++){
		int v=adj[u][i];
		int x=dfs2(v,mx).first;
		int y=dfs2(v,mx).second;
		if(x==-1){
			continue;
		}
		if(y>b){
			a=x;
			b=y;
		}
	}
	return {a,b};
}

int main(){
	ios::sync_with_stdio(0);
	cin.tie(0);
	int n,m;
	cin >> n >> m;
	int par[n+1];
	iota(par,par+n+1,0);
	pair<int,int> a[n+1];
	for(int i=1;i<=n;i++){
		cin >> a[i].first >> a[i].second;
	}	
	for(int i=1;i<=n;i++){
		for(int j=i+1;j<=n;j++){
			if(a[j].second>a[i].second){
				break;
			}
			else{
				par[j]=i;
			}
		}
	}
	for(int i=1;i<=n;i++){
		if(i==par[i]){
			continue;
		}
		else{
			adj[par[i]].push_back(i);
		}
	}
	for(int i=0;i<m;i++){
		int x;
		cin >> x;
		co[x]=1;
	}
	vector<int> ans;
	for(int i=1;i<=n;i++){
		if(par[i]==i){
			dfs(i);
			if(co[i]==0){
				continue;
			}
			else{
				ans.push_back(dfs2(i,co[i]).first);
			}
		}
	}
	cout << ans.size() << endl;
	sort(ans.begin(),ans.end());
	for(int i=0;i<ans.size();i++){
		cout << ans[i] << " ";
	}
}
